test(client): examples lint config

jason-ha opened the pull request:
to expect non-internal exports

Use /legacy for most imports, but some /beta and /alpha.
In example-utils modelLoader, use tinylicious-driver /test-utils. (example-utils readme notes it has non-production code.)

Copied implementation of `delay` from `core-utils` to avoid `/internal` import.

Skipped some packages per internal API use:
- apps/presence-tracker (pending rewrite in #23020)
- benchmarks/odspsnapshotfetch-perftestapp
- utils/bundle-size-tests
- utils/webpack-fluid-loader

There are a number of /internal imports that are suppressed.
packages with imported internals:
 - cell (x3) [AB#26903](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26903)
 - container-loader (x1) [AB#26907](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26907)
 - core-interfaces (x1) [AB#26908](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26908)
 - fluid-static (x1) [AB#26986](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26986)
 - local-driver (x3) [AB#26987](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26987)
 - merge-tree (x3) [AB#26905](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26905)
 - sequence (x8) [AB#26904](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26904)
 - test-runtime-utils (x2) [AB#26985](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26985)
 - test-utils (x7) [AB#26906](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/26906)

jason-ha committed:
test(client): examples lint config
to expect non-internal exports

Use /legacy for most imports, but some /beta and /alpha.
In example-utils modelLoader, use tinylicious-driver /test-utils. (example-utils readme notes it has non-production code.)

Skipped some packages per internal API use:
- apps/presence-tracker (pending rewrite in #23020)
- benchmarks/odspsnapshotfetch-perftestapp
- utils/bundle-size-tests
- utils/webpack-fluid-loader

There are a number of /internal imports that are suppressed.
packages with imported internals:
 - cell (x3) AB#26903
 - container-loader (x1) AB#26907
 - core-interfaces (x1) AB#26908
 - merge-tree (x2) AB#26905
 - sequence (x7) AB#26904
 - test-utils (x2) AB#26906

alexvy86 commented on examples/version-migration/same-container/src/modelVersion1/inventoryList.ts:
```diff
@@ -4,9 +4,10 @@
 import { EventEmitter } from "@fluid-example/example-utils";
-import { DataObject, DataObjectFactory } from "@fluidframework/aqueduct/internal";
+import { DataObject, DataObjectFactory } from "@fluidframework/aqueduct/legacy";
+// eslint-disable-next-line import/no-internal-modules -- #26903: `cell` internals used in examples
```
Just noting that as an example of migration, something we would expect external consumers would need to do, this use of internal/ APIs feels particularly egregious.

ChumpChief replied:
FWIW, I was using "cell" before it became unpopular :-P

In some future update we can change over to a different DDS, the specific choice of DDS isn't core to the message of the example.

alexvy86 reviewed (approved):
Net win overall IMO.

Josmithr commented on examples/.eslintrc.cjs:
```diff
@@ -0,0 +1,36 @@
+			{
+				allow: [
+					// Allow import of Fluid Framework external API exports.
+					"@fluidframework/*/{beta,alpha,legacy}",
```
Love this! So excited to have this in place!

Josmithr commented on examples/.eslintrc.cjs:
```diff
@@ -0,0 +1,36 @@
+		"import/no-internal-modules": [
+			"error",
+			{
+				allow: [
```
Nit: if we exported this array as a separate export, then configs that need to extend this list could do so explicitly, rather than re-declaring the necessary components.

jason-ha replied:
I like that idea. Need to find the proper CJS syntax... (it always confuses me a little)

jason-ha replied:
@Josmithr, I tried looking this up, but I don't know how to do it.

jason-ha replied:
@Josmithr, I tried looking this up, but I don't know how to do it.

Josmithr replied:
I think you should be able to just add an entry to `module.exports`. Since this isn't typed, ESLint should just ignore properties it doesn't understand. E.g.:

```js
const internalModulesAllow = [...];

module.exports = {
	rules: {...},
	internalModulesAllow,
}
```

Then the consuming package-level configs that care about that rule can just import the variable as normal:

```js
const { internalModulesAllow } = require(...);
```

Josmithr replied:
Ideally, we should be moving to ESLint 9 soon, given that ESLint 8 has now been out of support for a few months... ESLint 9 has ESM support (and TS support forthecoming). That would clean a lot of this up, IMO. But it's going to be a big project...

jason-ha replied:
Ah, so just let it pollute the lint config. I think we let it wait until there is a use case.

jason-ha replied:
I tried this as I found a use, but alas we get a "no" from tooling:
```text
@fluid-example/webflow: Error: ESLint configuration in .eslintrc.cjs » ../../.eslintrc.cjs is invalid:
@fluid-example/webflow:         - Unexpected top-level property "importInternalModulesAllowed".
@fluid-example/webflow: 
@fluid-example/webflow: Referenced from: /home/jasonha/ff/FluidFramework/examples/data-objects/webflow/.eslintrc.cjs
```
We can work around though it takes mucking with all extends (AFAIK).

jason-ha replied:
I looked at the code some more and found more /internal uses in places with missing lint coverage on test code and alterations preventing detection. Fixed all that and implemented a way to share the allow array.

Josmithr commented on examples/.eslintrc.cjs:
```diff
@@ -0,0 +1,36 @@
+ * Licensed under the MIT License.
+ */
+
+module.exports = {
```
Oh, I see. The individual example configs extend both. No strong opinion here, but it seems like it would be a bit simpler / slightly less error prone to just have this extend the base config, and have the individual packages just extend this.

jason-ha replied:
I think the problem with this having a base is that the packages use different fluid configs. So if this were to have a base, then they would extend from this first and then they would override with the one they really want. But that would set these import rules to whatever is in the latest listed config.\
Unless we do programmatic generation, this appears to be the simplest config.
I am adding a comment to explain use.

Josmithr replied:
Good point. Programmatic config generation is something I'd like to look into in the future. We have a lot of variable axis for our packages that would be interesting to capture that way. But we should migrate to ESLint 9 first.

Josmithr commented on examples/utils/webpack-fluid-loader/.eslintrc.cjs:
```diff
@@ -7,6 +7,8 @@ module.exports = {
 	extends: [
 		require.resolve("@fluidframework/eslint-config-fluid/minimal-deprecated"),
 		"prettier",
+		// There are a lot of internal APIs leveraged here.
+		// "../../.eslintrc.cjs",
```
We should also probably discuss whether or not we actually want to be publishing this. As it is currently written, I don't think this is an "example". It's also so finely tuned to our test scenarios, that it really shouldn't be a public utility either; it should be a private implementation detail of our "examples" (which I agree with what others have suggested - we should call most of these "test apps" not "examples".

jason-ha replied:
At the same time, why shouldn't the external API support what is happening here? (I didn't look into the details. Maybe more of its tricks are above board and not interesting to external parties.)

alexvy86 replied:
Actually, I think technically they do, and if you connect them both to an external service things will work. The point of webpack-fluid-loader as I understand it was being able to showcase collaboration _while using local server_, which lives in-memory (i.e. the browser tab) so it doesn't support collaboration across browser tabs. And this package was concocted to take an app that makes no special assumptions about how it is being rendered, and renders it twice in the same tab, where both instances can be connected to the local server. I imagine the internal/-ness in this package has more to do with the fact that we want to use local server, than with rendering an app twice in the same tab.

jason-ha replied:
Want to make sure we are on the same page about questions/statements.
I am saying that you can't write webpack-fluid-loader without using /internal APIs IFAIK. If the webpack-fluid-loader scenarios are interesting outside of FF repo, then either webpack-fluid-loader should be published to enable these scenarios OR the APIs need exposed to let someone write a thing like webpack-fluid-loader.

If there isn't desire for that, then such things would not be high priority.
Probably there should be a pass over this package to isolate those true /internal APIs and allow just those. That will reveal the special API surface.

ChumpChief replied:
IMO the long-term goal should be to eliminate webpack-fluid-loader.  It's not representative of how customers use Fluid.  Basically everything outside of the /examples/data-objects directory is the product of slow conversion away from WFL.

The main blockers for completing the transition are:
1. We don't have ODSP support in other examples currently, but should (i.e. `start:spo` script).  Same is true for r11s and docker but I don't think we really care about those as much.
2. There are a variety of utility functionalities it provides that may or may not be useful at this point - we should evaluate whether we care about these (e.g. the extra routes like /pushLogin, /odspLogin, /file*, magic URLs like `#prefetch`, `#binarySnapshot`).  However if useful I suspect these functionalities would be better supported by a standalone dedicated package rather than tying them into the webpack dev server for the example.
3. Labor.  It's not particularly hard to convert from WFL to non-WFL, but it takes some time and testing.

jason-ha commented on examples/data-objects/clicker/src/agent.ts:
```diff
@@ -3,8 +3,9 @@
  * Licensed under the MIT License.
  */
 
+// eslint-disable-next-line import/no-internal-modules -- #26908: `core-interfaces` internal used in examples
```
Same hope and now we know.
I think I mostly agree. Until we have a robust FluidExamples set and tight (fast and required) integration pipeline for testing we need at least some of these. We'd still want the "test apps" to limit themselves to non-internals. Definitely forward progress.

jason-ha commented on examples/utils/bundle-size-tests/.eslintrc.cjs:
```diff
@@ -7,6 +7,8 @@ module.exports = {
 	extends: [
 		require.resolve("@fluidframework/eslint-config-fluid/minimal-deprecated"),
 		"prettier",
+		// There are a lot of intentional internal APIs leveraged here for simplicity. Skip common example rules:
+		// "../../.eslintrc.cjs",
```
I don't like tools though for this. We can create dedicated test space. If we renamed examples to tests, this would fit into place but still have internal use at odds with other tests.

alexvy86 replied:
It's not really about tests either though, it's just a tool we use in our calculation of bundle sizes. It doesn't _test_ anything.

jason-ha committed:
docs(client-examples): note lint config setup pattern

jason-ha committed:
fix(client-examples): build breaks from /legacy use

jason-ha committed:
test(client): examples lint config (more)
to use non-internal exports

Refactor examples/.eslintrc.cjs to have allow lists shareable via .eslintrc.data.cjs.
Clean up more packages to use new config and eliminate more /internal imports.

jason-ha committed:
docs(client): update work item ids

jason-ha committed:
merge: 'main' into examples/reset-imports-to-external-use

jason-ha committed:
docs(client): correct work item ids

Josmithr commented on examples/.eslintrc.data.cjs:
```diff
@@ -0,0 +1,71 @@
+ * Licensed under the MIT License.
+ */
+
+// To facilitate reuse of configuration elements this module exports some
```
I like this pattern!

Josmithr commented on examples/external-data/tests/utilities.ts:
```diff
@@ -19,3 +19,11 @@ export async function closeServer(server: Server): Promise<void> {
+ * @param timeMs - Time in milliseconds to wait.
+ * @internal
+ */
+export const delay = async (timeMs: number): Promise<void> =>
```
Nit: should this go in the `example-utils` package (to avoid re-implementing this behavior in too many places)?

jason-ha replied:
Maybe. It has a bit of a code smell. My guess is that there are better patterns.

ChumpChief replied:
IMO example-utils should be small and its members should not be long-lived.  Code duplication is actually a benefit in the examples - customers are best able to consume an example if they can avoid cross referencing between packages.

The current members of the example-utils package are mostly intended to be "temporary" (though many have been there for a long time).  The goal is to experiment with a pattern broadly across multiple examples and then promote that pattern to a real shipping package that customers can use.  migration-tools is one example that finally has its own package, some of the other functionalities might end up in aqueduct, etc.

Josmithr reviewed (approved):
Looks great! I like the new config pattern. And thanks for filing all of those follow-up items!

Josmithr commented:
Not for this PR, but for the example apps built on our "3P" API, should we further restrict the imports to not include `@legacy`?

jason-ha commented on examples/.eslintrc.data.cjs:
```diff
@@ -0,0 +1,71 @@
+};
+
+module.exports = {
+	importInternalModulesAllowed: importInternalModulesAllowed,
+	importInternalModulesAllowedForTest: importInternalModulesAllowedForTest,
+	lintConfig: lintConfig,
```
Not in CommonJS. I tried that to start and had to expand.

jason-ha commented:
> Not for this PR, but for the example apps built on our "3P" API, should we further restrict the imports to not include `@legacy`?

presence-tracker revision might end up on "3P", but we don't have many that aren't using /legacy. May quick check says there are 13 that might be okay. More than I thought actually.
Probably we'll want a .eslintrc.legacy.cjs.
